esmvalcore/preprocessor/_time.py

```diff
@@ -1292,3 +1298,381 @@ def resample_time(
+    <https://numpy.org/doc/stable/user/basics.indexing.html#advanced-indexing>`__
+    for the time dimension of `data` with shape (time, lon). It is used to
+    reorder the data along the time axis based on the longitude axis.
+
+    """
+    # Apart from the time index, all other dimensions will stay the same; this
+    # is ensured with np.ogrid
+    idx = np.ogrid[tuple(slice(0, d) for d in data.shape)]
+    idx[-2] = np.broadcast_to(time_index, data.shape)  # '-2' is always time
+    new_data = data[tuple(idx)]
+    return new_data
+
+
+def _transform_to_lst_lazy(
+    data: da.core.Array,
+    *,
+    time_index: np.ndarray,
+    mask: np.ndarray,
+    time_dim: int,
+    lon_dim: int,
+    output_dtypes: DTypeLike,
+) -> da.core.Array:
+    """Transform array with UTC coord to local solar time (LST) coord.
+
+    Note
+    ----
+    This function is the lazy version of `_transform_to_lst_eager` using
+    dask's :func:`dask.array.apply_gufunc`.
+
+    `data` needs to be at least 2D. `time_dim` and `lon_dim` correspond to the
+    dimensions that describe time and longitude dimensions in `data`,
+    respectively.
+
+    `time_index` is an `advanced index
+    <https://numpy.org/doc/stable/user/basics.indexing.html#advanced-indexing>`__
+    for the time dimension of `data` with shape (time, lon). It is used to
+    reorder the data along the time axis based on the longitude axis.
+
+    `mask` is 2D with shape (time, lon) that will be applied to the final data.
+
+    """
+    new_data = da.apply_gufunc(
+        _transform_to_lst_eager,
+        '(t,y)->(t,y)',
+        data,
+        axes=[(time_dim, lon_dim), (time_dim, lon_dim)],
+        output_dtypes=output_dtypes,
+        time_index=time_index,
+    )
+
+    # Apply properly broadcasted mask
+    new_mask = broadcast_to_shape(
+        da.from_array(mask), new_data.shape, (time_dim, lon_dim)
+    )
+    new_mask = new_mask.rechunk(new_data.chunksize)
+    new_mask |= da.ma.getmaskarray(new_data)
+
+    return da.ma.masked_array(new_data, mask=new_mask)
+
+
+def _transform_cube_to_lst(cube: Cube) -> Cube:
+    """Transform cube to local solar time (LST) coordinate (lazy; in-place)."""
+    # Rechunk cube properly (it must not be chunked along time and longitude
+    # dimension); this also creates a new cube so the original input cube is
+    # not overwritten
+    complete_coords = [
+        cube.coord('time', dim_coords=True), cube.coord('longitude'),
+    ]
+    cube = rechunk_cube(cube, complete_coords)
+
+    time_coord = cube.coord('time', dim_coords=True)
+    lon_coord = cube.coord('longitude')
+    time_dim = cube.coord_dims(time_coord)[0]
+    lon_dim = cube.coord_dims(lon_coord)[0]
+
+    # Transform cube data
+    (time_index, mask) = _get_time_index_and_mask(time_coord, lon_coord)
+    _transform_arr = partial(
+        _transform_to_lst_lazy,
+        time_index=time_index,
+        mask=mask,
+    )
+    cube.data = _transform_arr(
+        cube.lazy_data(),
```

Comment:
Good point. I also though about this, but unfortunately it's not as simple as using `core_data()` instead of `lazy_data()` here, since `da.apply_gufunc` will always return a dask array. In addition, we need to also consider other dimensional metadata here. For example, a cube can have realized data, but lazy cell measures.

I'll try to think of a solution that properly handles all cases. 

One question: how should `rechunk_cube` deal with numpy arrays? Leave them like this? Or transform them to a dask array?

Reply:
All right, implemented this in https://github.com/ESMValGroup/ESMValCore/pull/2258/commits/b77f6b95f7f789d70dc17026c0d2a07602a65de0. `local_solar_time` will only return dask arrays if the input is already lazy.

`rechunk_cube` now never touches numpy arrays (https://github.com/ESMValGroup/ESMValCore/pull/2258/commits/9a2fa74d570eda3cf5cb63b487594f21f5ccea96).

Reply:
> There is, however, the question of Numpy lazy data but I am not too familiar with that yet

I haven't heard of that before, do you have a reference? The only thing resembling lazy data in numpy that I know of are [views](https://numpy.org/doc/stable/user/basics.copies.html#view), some numpy functions will return those instead of computing a new array.

Reply:
was thinking of `mmap()` - here's a cool side-by-side of `mmap` vs HDF5 vs Zarr https://pythonspeed.com/articles/mmap-vs-zarr-hdf5/